### PR TITLE
Feature: Create Slack-formatted slack report

### PR DESCRIPTION
**Pull request by royrapoport**

### CODE REVIEW

File: report.py [hunk 1]
```diff
@@ -210,6 +211,9 @@ def accum_timestats(self, message):
             tz_offset = config.default_tz_offset
         tz = user.get("tz", "Unknown")
         self.increment(["timezone", tz], message)
+        if tz not in self._data['posters_per_timezone']:
+            self._data['posters_per_timezone'][tz] = {}
+        self._data['posters_per_timezone'][tz][uid] = 1
```
Comment (jdq22) [inline report.py:214]: You could use collections.defaultdict instead.

File: slack_global_report.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+import enricher
+import slack_token
+import slack_formatter
+
+
+class SlackGlobalReport(object):
+
+    def __init__(self, fake=False):
+        self.sf = slack_formatter.SlackFormatter()
+        random.seed(time.time())
```
Comment (jdq22) [inline slack_global_report.py]: What is this for?

File: slack_global_report.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        enricher.Enricher(fake=self.fake).enrich(ur)
+        enricher.Enricher(fake=self.fake).enrich(previous)
+        blocks = self.make_report(ur, previous)
+        if not send:
+            print("Saving report to slack.json")
+            f = open("slack.json", "w")
+            f.write(json.dumps(blocks, indent=4))
+            f.close()
+        # If set to true, this message will be sent as the user who owns the token we use
+        as_user = False
```
Comment (jdq22) [inline slack_global_report.py:207]: Is it possible to send this as anyone besides Slack API Tester and you?

File: slack_global_report.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    def make_header(self, ur, pur):
+        blocks = []
+        header = "*Slack Activity Report for the Week Between {} and {}*"
+        header = header.format(ur['start_date'], ur['end_date'])
+        blocks.append(self.sf.text_block(header))
+        stats = ur['statistics']
+        posters = int(stats['posters'])
+        active_users = stats['active_users']
+        percent = (posters * 100.0) / active_users
+        text = "*{:,}/{:,}* (or *{:.1f}%*) users posted messages\n".format(posters, active_users, percent)
```
Comment (jdq22) [inline slack_global_report.py]: Before: "828/12,937 (or 6.4%) users posted messages"

Can you change it to: "828/12,937 (or 6.4% of) users posted messages"

File: slack_global_report.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        text += "Median message count was *{}*\n".format(stats['median messages'])
+        text += "The top ten posters contributed *{:.1f}%* of all messages (lower is better)\n".format(stats['topten messages'])
```
Comment (jdq22) [inline slack_global_report.py]: Can you bold "ten" so it matches the rest of the other bolded stats?

File: slack_global_report.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        text += "The top *{}* posters (higher is better) accounted for about 50% of total volume\n".format(stats['50percent of words'])
+        blocks.append(self.sf.text_block(text))
+        w = stats['words']
+        m = stats['messages']
+        # it = interim text
+        it = "People posted *{:,}* words in *{:,}* messages (approximately *{:,}* pages), "
```
Comment (jdq22) [inline slack_global_report.py:46]: What defines a page?

File: slack_global_report.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+                it += " (new)"
+            it += "{:,} members ".format(ci['members'])
+
+            m = channels[channel][0]
+            w = channels[channel][1]
+            p = len(cu)
+            it += "*{:,}* posters *{:,}* words *{:,}* messages ".format(p, w, m)
+            it += "*{:.1f}* words/poster ".format(w/p)
+            it += "*{:.1f}%* of total traffic, ".format(cs['percent'])
+            it += "*{:.1f}%* cumulative of total ".format(cs['cpercent'])
```
Comment (jdq22) [inline slack_global_report.py:80]: Before: "#cooking 426 members 66 posters 8,456 words 451 messages 128.1 words/poster 3.3% of total traffic, 16.0% cumulative of total"

Can you add commas so it looks like this?
"#cooking: 426 members, 66 posters, 8,456 words, 451 messages, 128.1 words/poster, 3.3% of total traffic, 16.0% cumulative of total"

File: slack_global_report.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+            w = usu['count'][1]
+            per = usu['percent_of_words']
+            cper = usu['cum_percent_of_words']
+            rphw = usu['reactions'] * 100.0 / w
+            w_per_m = w / m
+            t = usu['thread_messages']
+            it = "{}. *{}* ".format(idx + 1, self.sf.show_uid(uid))
+            it += "*{:,}* words *{:,}* messages *{:.1f}* w/m ".format(w, m, w_per_m)
+            it += "*{:.1f}* rphw ".format(rphw)
+            it += "*{:.1f}%* *{:.1f}%* cumulative\n".format(per, cper)
```
Comment (jdq22) [inline slack_global_report.py]: Similar comma request.

File: slack_global_report.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        stats = ur['statistics']
+        us = ur['user_stats']
+        uids = stats['50percent users for words'][:top]
+        for idx, uid in enumerate(uids):
+            usu = us[uid]
+            m = usu['count'][0]
+            w = usu['count'][1]
+            per = usu['percent_of_words']
+            cper = usu['cum_percent_of_words']
+            rphw = usu['reactions'] * 100.0 / w
```
Comment (jdq22) [inline slack_global_report.py:100]: Can you add some copy at the top of this section that explains what rphw is?

File: slack_global_report.py (hunk 8 above)
Comment (jdq22) [inline slack_global_report.py]: Is this cumulative of total? If so, can you change it to be cumulative of total?

File: report.py (hunk 1 above)
Comment (royrapoport) [reply report.py:214]: Yup. Done.

File: slack_global_report.py (hunk 2 above)
Comment (royrapoport) [reply slack_global_report.py]: Some old code.  Might need to resurrect at some point -- goal is to be able to basically create fake activity reports so we can show people what this does without violating privacy, but I'm not going to worry about it for now.  Clean it up.

File: slack_global_report.py (hunk 3 above)
Comment (royrapoport) [reply slack_global_report.py:207]: Yeah, it's whoever has the Slack API key.  Nothing for us to do here.

File: slack_global_report.py (hunk 4 above)
Comment (royrapoport) [reply slack_global_report.py]: Done

File: slack_global_report.py (hunk 5 above)
Comment (royrapoport) [reply slack_global_report.py]: Done

File: slack_global_report.py (hunk 6 above)
Comment (royrapoport) [reply slack_global_report.py:46]: Added a code comment -- we approximate 500 words per page

File: slack_global_report.py (hunk 7 above)
Comment (royrapoport) [reply slack_global_report.py:80]: Done

File: slack_global_report.py (hunk 8 above)
Comment (royrapoport) [reply slack_global_report.py]: Done

File: slack_global_report.py (hunk 9 above)
Comment (royrapoport) [reply slack_global_report.py:100]: Added

File: slack_global_report.py (hunk 8 above)
Comment (royrapoport) [reply slack_global_report.py]: Done

File: slack_global_report.py (hunk 6 above)
Comment (jdq22) [reply slack_global_report.py:46]: Can you add a note somewhere in the report that explains what a page is? I know other people will ask :)

File: slack_global_report.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+            w = match[1]
+            it = "*{}* *{:,}* words in *{:,}* messages\n".format(day, w, m)
+            text += it
+        blocks.append(self.sf.text_block(text))
+        blocks.append(self.sf.divider())
+        return blocks
+
+    def hours(self, ur, pur):
+        blocks = []
+        header = "*Activity Per Hour (on Weeekdays)*"
```
Comment (jdq22) [inline slack_global_report.py]: Typo

Comment (royrapoport) [reply slack_global_report.py]: fixed

File: slack_global_report.py (hunk 6 above)
Comment (royrapoport) [reply slack_global_report.py:46]: Done

**Review by jdq22 — APPROVED** [review]
This is awesome. Great work.

**Pull request merged**